File: web/src/services/session-manager.ts
/**
 * Session Manager - Session Lifecycle Management
 *
 * Manages session creation, execution, and state tracking.
 * Supports concurrent session execution with JSON file persistence.
 */

import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs'
import { join, dirname } from 'path'
import type { Session, SessionStatus, SessionSummary, Market, Model, AnalyzeRequest } from '../types'
import { DATA_DIR, AGENT_ORDER } from '../config'
import { sseManager } from './sse-manager'
import { analysisClient } from './analysis-client'

// Session storage file path
const SESSIONS_FILE = join(DATA_DIR, 'sessions.json')

// Generate short ID
function generateId(): string {
  return Math.random().toString(36).substring(2, 10)
}

// Get next agent
function getNextAgent(currentAgent: string): string {
  const idx = AGENT_ORDER.indexOf(currentAgent)
  if (idx >= 0 && idx + 1 < AGENT_ORDER.length) {
    return AGENT_ORDER[idx + 1]
  }
  return ''
}

export class SessionManager {
  private sessions: Map<string, Session> = new Map()
  private runningTasks: Map<string, AbortController> = new Map()

  /**
   * Load sessions from file
   */
  loadSessions(): void {
    if (!existsSync(SESSIONS_FILE)) {
      return
    }

    try {
      const data = JSON.parse(readFileSync(SESSIONS_FILE, 'utf-8'))
      for (const sessionData of data) {
        const session = this.parseSession(sessionData)
        // If previously running, mark as error (server restarted)
        if (session.status === 'running') {
          session.status = 'error'
          session.errorMsg = 'Server restarted, analysis interrupted'
          session.currentAgent = ''
        }
        this.sessions.set(session.id, session)
      }
      console.log(`Loaded ${this.sessions.size} sessions from disk`)
    } catch (e) {
      console.error('Failed to load sessions:', e)
    }
  }

  /**
   * Save sessions to file
   */
  private saveSessions(): void {
    try {
      // Ensure directory exists
      const dir = dirname(SESSIONS_FILE)
      if (!existsSync(dir)) {
        mkdirSync(dir, { recursive: true })
      }

      const data = Array.from(this.sessions.values())
      writeFileSync(SESSIONS_FILE, JSON.stringify(data, null, 2), 'utf-8')
    } catch (e) {
      console.error('Failed to save sessions:', e)
    }
  }

  /**
   * Parse session data
   */
  private parseSession(data: any): Session {
    return {
      id: data.id || generateId(),
      ticker: data.ticker || '',
      stockName: data.stock_name || data.stockName || '',
      market: data.market || 'A-share',
      model: data.model || 'deepseek-v3',
      startDate: data.start_date || data.startDate || '',
      endDate: data.end_date || data.endDate || '',
      userId: data.user_id || data.userId || '',
      status: data.status || 'pending',
      decision: data.decision || '',
      progress: data.progress || [],
      reports: data.reports || {},
      createdAt: data.created_at || data.createdAt || new Date().toISOString(),
      errorMsg: data.error_msg || data.errorMsg || '',
      currentAgent: data.current_agent || data.currentAgent || '',
    }
  }

  /**
   * Create a new session
   */
  createSession(params: {
    ticker: string
    stockName: string
    market: Market
    model: Model
    startDate: string
    endDate: string
    userId: string
  }): Session {
    const session: Session = {
      id: generateId(),
      ticker: params.ticker,
      stockName: params.stockName,
      market: params.market,
      model: params.model,
      startDate: params.startDate,
      endDate: params.endDate,
      userId: params.userId,
      status: 'pending',
      decision: '',
      progress: [],
      reports: {},
      createdAt: new Date().toISOString().replace('T', ' ').substring(0, 19),
      errorMsg: '',
      currentAgent: '',
    }

    this.sessions.set(session.id, session)
    this.saveSessions()

    return session
  }

  /**
   * Get a session
   */
  getSession(sessionId: string, userId?: string, isAdmin?: boolean): Session | null {
    const session = this.sessions.get(sessionId)
    if (!session) {
      return null
    }

    // Admin can access all sessions
    if (isAdmin) {
      return session
    }

    // Regular users can only access their own sessions
    if (userId && session.userId && session.userId !== userId) {
      return null
    }

    return session
  }

  /**
   * Get all sessions (sorted by creation time descending)
   */
  getAllSessions(userId?: string, isAdmin?: boolean): Session[] {
    let sessions: Session[]

    if (isAdmin) {
      sessions = Array.from(this.sessions.values())
    } else if (userId) {
      sessions = Array.from(this.sessions.values()).filter(s => s.userId === userId)
    } else {
      sessions = Array.from(this.sessions.values()).filter(s => !s.userId)
    }

    return sessions.sort((a, b) =>
      new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime()
    )
  }

  /**
   * Get all session summaries (without reports, optimized for list loading)
   */
  getAllSessionSummaries(userId?: string, isAdmin?: boolean): SessionSummary[] {
    const sessions = this.getAllSessions(userId, isAdmin)

    // Calculate queue position per user
    const userQueuePositions = new Map<string, number>()

    return sessions.map(session => {
      const summary: SessionSummary = {
        id: session.id,
        ticker: session.ticker,
        stockName: session.stockName,
        market: session.market,
        model: session.model,
        startDate: session.startDate,
        endDate: session.endDate,
        userId: session.userId,
        status: session.status,
        decision: session.decision,
        progress: session.progress,
        createdAt: session.createdAt,
        errorMsg: session.errorMsg,
        currentAgent: session.currentAgent,
      }

      // Calculate queue position
      if (session.status === 'queued') {
        const key = session.userId || '__anonymous__'
        const pos = (userQueuePositions.get(key) || 0) + 1
        userQueuePositions.set(key, pos)
        summary.queuePosition = pos
      }

      return summary
    })
  }

  /**
   * Get running session count
   */
  getRunningCount(userId?: string, isAdmin?: boolean): number {
    let count = 0
    for (const session of this.sessions.values()) {
      if (session.status !== 'running') continue

      if (isAdmin) {
        count++
      } else if (userId && session.userId === userId) {
        count++
      } else if (!userId && !session.userId) {
        count++
      }
    }
    return count
  }

  /**
   * Check if user has a running session
   * Used to limit one concurrent analysis task per user
   */
  getUserRunningSession(userId: string): Session | null {
    for (const session of this.sessions.values()) {
      if (session.status === 'running' && session.userId === userId) {
        return session
      }
    }
    return null
  }

  /**
   * Delete a session
   */
  deleteSession(sessionId: string): boolean {
    if (this.sessions.has(sessionId)) {
      // Cancel running task
      const controller = this.runningTasks.get(sessionId)
      if (controller) {
        controller.abort()
        this.runningTasks.delete(sessionId)
      }

      this.sessions.delete(sessionId)
      this.saveSessions()
      return true
    }
    return false
  }

  /**
   * Start session execution (supports queuing)
   */
  async startSession(sessionId: string, analysts: string[]): Promise<void> {
    const session = this.sessions.get(sessionId)
    if (!session) {
      return
    }

    // Save analysts config for later queued start
    ;(session as any)._analysts = analysts

    // Check if user has a running task
    const runningSession = this.getUserRunningSession(session.userId)
    if (runningSession) {
      // Has running task, add to queue
      session.status = 'queued'
      this.saveSessions()
      sseManager.publish(sessionId, session)
      console.log(`[Session ${sessionId}] Queued (user ${session.userId} has running session ${runningSession.id})`)
      return
    }

    // No running task, start immediately
    this.startSessionNow(sessionId, analysts)
  }

  /**
   * Start session immediately (internal method, skips queue check)
   */
  private startSessionNow(sessionId: string, analysts: string[]): void {
    const session = this.sessions.get(sessionId)
    if (!session) {
      return
    }

    // Update status to running
    session.status = 'running'
    session.currentAgent = AGENT_ORDER[0] || ''
    this.saveSessions()
    sseManager.publish(sessionId, session)
    console.log(`[Session ${sessionId}] Started`)

    // Create AbortController for cancellation
    const controller = new AbortController()
    this.runningTasks.set(sessionId, controller)

    // Run analysis in background
    this.runAnalysis(sessionId, analysts, controller.signal).catch(console.error)
  }

  /**
   * Start next queued session for this user
   */
  private startNextQueuedSession(userId: string): void {
    // Find the earliest created queued session for this user
    let nextSession: Session | null = null
    let earliestTime = Infinity

    for (const session of this.sessions.values()) {
      if (session.status === 'queued' && session.userId === userId) {
        const time = new Date(session.createdAt).getTime()
        if (time < earliestTime) {
          earliestTime = time
          nextSession = session
        }
      }
    }

    if (nextSession) {
      const analysts = (nextSession as any)._analysts || ['market', 'social', 'news', 'fundamentals']
      console.log(`[Queue] Starting next session ${nextSession.id} for user ${userId}`)
      this.startSessionNow(nextSession.id, analysts)
    }
  }

  /**
   * Run analysis (in background) - uses token-level streaming
   */
  private async runAnalysis(
    sessionId: string,
    analysts: string[],
    signal: AbortSignal
  ): Promise<void> {
    const session = this.sessions.get(sessionId)
    if (!session) {
      return
    }

    try {
      const request: AnalyzeRequest = {
        ticker: session.ticker,
        date: session.endDate,
        market: session.market,
        analysts,
        model: session.model,
      }

      // Track current agent's accumulated content
      let currentAgent = ''
      let accumulatedContent = ''
      let tokenCount = 0

      for await (const chunk of analysisClient.streamTokens(request)) {
        if (signal.aborted) return

        const { type, agent, content } = chunk

        switch (type) {
          case 'heartbeat':
            // Heartbeat message, ignore
            break

          case 'node_start':
            // Agent started processing
            if (agent) {
              currentAgent = agent
              accumulatedContent = ''
              tokenCount = 0
              session.currentAgent = agent
              this.saveSessions()
              sseManager.publish(sessionId, session)
            }
            break

          case 'token':
            // Received token, accumulate and push in real-time
            if (agent && content) {
              accumulatedContent += content
              tokenCount++
              session.reports[agent] = accumulatedContent

              // Push every token for smooth typewriter effect
              sseManager.publish(sessionId, session)
            }
            break

          case 'node_end':
            // Agent completed
            if (agent && content) {
              session.reports[agent] = content
              session.progress.push(agent)
              session.currentAgent = getNextAgent(agent)
              this.saveSessions()
              sseManager.publish(sessionId, session)
            }
            currentAgent = ''
            accumulatedContent = ''
            break

          case 'complete':
            // Analysis complete
            session.decision = (content || 'HOLD') as any
            session.status = 'completed'
            session.currentAgent = ''
            this.saveSessions()
            sseManager.publish(sessionId, session)
            return

          case 'quota_error':
            session.status = 'error'
            session.errorMsg = 'API quota exhausted'
            session.currentAgent = ''
            this.saveSessions()
            sseManager.publish(sessionId, session)
            return

          case 'timeout_error':
            session.status = 'error'
            session.errorMsg = 'Network response timeout'
            session.currentAgent = ''
            this.saveSessions()
            sseManager.publish(sessionId, session)
            return

          case 'error':
            session.status = 'error'
            session.errorMsg = (content || 'Unknown error').substring(0, 200)
            session.currentAgent = ''
            this.saveSessions()
            sseManager.publish(sessionId, session)
            return
        }
      }

      // If stream ended without a complete event, mark as completed
      if (session.status === 'running') {
        session.status = 'completed'
        session.currentAgent = ''
        this.saveSessions()
        sseManager.publish(sessionId, session)
      }

    } catch (e) {
      const error = e as Error
      console.error(`[Session ${sessionId}] Analysis error:`, error.message)
      session.status = 'error'
      session.errorMsg = error.message?.substring(0, 200) || 'Unknown error'
      session.currentAgent = ''
      this.saveSessions()
      sseManager.publish(sessionId, session)
    } finally {
      this.runningTasks.delete(sessionId)

      // Start next queued task for this user
      if (session.userId) {
        this.startNextQueuedSession(session.userId)
      }
    }
  }

  /**
   * Retry a failed session
   */
  async retrySession(sessionId: string, analysts?: string[]): Promise<void> {
    const session = this.sessions.get(sessionId)
    if (!session) {
      return
    }

    // Reset state
    session.status = 'pending'
    session.errorMsg = ''
    session.progress = []
    session.reports = {}
    session.decision = ''
    session.currentAgent = ''
    this.saveSessions()

    // Restart
    await this.startSession(
      sessionId,
      analysts || ['market', 'social', 'news', 'fundamentals']
    )
  }

  /**
   * Clear completed sessions
   */
  clearCompleted(): number {
    const toDelete: string[] = []

    for (const [id, session] of this.sessions) {
      if (session.status === 'completed' || session.status === 'error') {
        toDelete.push(id)
      }
    }

    for (const id of toDelete) {
      this.sessions.delete(id)
    }

    if (toDelete.length > 0) {
      this.saveSessions()
    }

    return toDelete.length
  }

  /**
   * Graceful shutdown: mark all running sessions as interrupted
   * Called before server shutdown
   */
  gracefulShutdown(): void {
    console.log('[SessionManager] Graceful shutdown initiated')
    let interrupted = 0

    for (const session of this.sessions.values()) {
      if (session.status === 'running') {
        session.status = 'error'
        session.errorMsg = 'Server restarted, analysis interrupted (click retry to continue)'
        session.currentAgent = ''
        interrupted++
      }
    }

    if (interrupted > 0) {
      this.saveSessions()
      console.log(`[SessionManager] Marked ${interrupted} running sessions as interrupted`)
    }

    // Cancel all running tasks
    for (const [sessionId, controller] of this.runningTasks) {
      controller.abort()
      console.log(`[SessionManager] Aborted task for session ${sessionId}`)
    }
    this.runningTasks.clear()
  }
}

// Global singleton
export const sessionManager = new SessionManager()

File: gui/session_manager.py
"""
Session 管理模块

管理分析 Session 的创建、执行、状态跟踪。
支持并发执行多个 Session，数据持久化到 JSON 文件。
"""

import json
import uuid
import threading
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional, Callable
from concurrent.futures import ThreadPoolExecutor
from enum import Enum

from .config import AGENT_ORDER

# SSE 发布器（由 server.py 设置，避免循环导入）
_sse_publisher = None


def set_sse_publisher(publisher):
    """设置 SSE 发布器（由 server.py 调用）"""
    global _sse_publisher
    _sse_publisher = publisher


# Session 存储文件路径
# Railway Volume 挂载到 /data，本地开发用项目目录
import os
_data_dir = os.environ.get("DATA_DIR", str(Path(__file__).parent.parent / "data"))
SESSIONS_FILE = Path(_data_dir) / "sessions.json"


class SessionStatus(Enum):
    """Session 状态枚举"""
    PENDING = "pending"
    RUNNING = "running"
    COMPLETED = "completed"
    ERROR = "error"


@dataclass
class Session:
    """Session 数据结构"""
    id: str
    ticker: str
    stock_name: str
    market: str
    start_date: str
    end_date: str
    user_id: str = ""  # 所属用户
    status: SessionStatus = SessionStatus.PENDING
    decision: str = ""
    progress: List[str] = field(default_factory=list)
    reports: Dict[str, str] = field(default_factory=dict)
    created_at: str = ""
    error_msg: str = ""
    current_agent: str = ""

    def __post_init__(self):
        if not self.created_at:
            self.created_at = datetime.now().strftime("%Y-%m-%d %H:%M:%S")

    def to_dict(self) -> dict:
        """转换为字典（用于 JSON 序列化）"""
        return {
            "id": self.id,
            "ticker": self.ticker,
            "stock_name": self.stock_name,
            "market": self.market,
            "start_date": self.start_date,
            "end_date": self.end_date,
            "user_id": self.user_id,
            "status": self.status.value,
            "decision": self.decision,
            "progress": self.progress,
            "reports": self.reports,
            "created_at": self.created_at,
            "error_msg": self.error_msg,
            "current_agent": self.current_agent,
        }

    @staticmethod
    def from_dict(data: dict) -> "Session":
        """从字典创建 Session"""
        return Session(
            id=data.get("id", str(uuid.uuid4())),
            ticker=data.get("ticker", ""),
            stock_name=data.get("stock_name", ""),
            market=data.get("market", "A-share"),
            start_date=data.get("start_date", ""),
            end_date=data.get("end_date", ""),
            user_id=data.get("user_id", ""),
            status=SessionStatus(data.get("status", "pending")),
            decision=data.get("decision", ""),
            progress=data.get("progress", []),
            reports=data.get("reports", {}),
            created_at=data.get("created_at", ""),
            error_msg=data.get("error_msg", ""),
            current_agent=data.get("current_agent", ""),
        )


class SessionManager:
    """
    Session 管理器

    负责 Session 的生命周期管理和并发执行。
    """

    def __init__(self, max_workers: int = 3):
        """
        初始化 Session 管理器

        Args:
            max_workers: 最大并发 Session 数
        """
        self._sessions: Dict[str, Session] = {}
        self._lock = threading.RLock()  # 使用可重入锁，避免嵌套调用死锁
        self._executor = ThreadPoolExecutor(max_workers=max_workers)
        self._callbacks: Dict[str, Callable] = {}  # session_id -> callback
        self._analysis_service = None

        # 从文件加载 sessions
        self._load_sessions()

    def _load_sessions(self):
        """从 JSON 文件加载 sessions"""
        if not SESSIONS_FILE.exists():
            return

        try:
            with open(SESSIONS_FILE, "r", encoding="utf-8") as f:
                data = json.load(f)

            for session_data in data:
                session = Session.from_dict(session_data)
                # 如果 session 之前是 running 状态，标记为 error（因为服务器重启了）
                if session.status == SessionStatus.RUNNING:
                    session.status = SessionStatus.ERROR
                    session.error_msg = "服务器重启，分析中断"
                    session.current_agent = ""
                self._sessions[session.id] = session
        except Exception as e:
            print(f"加载 sessions 失败: {e}")

    def _save_sessions(self):
        """保存 sessions 到 JSON 文件"""
        try:
            # 确保目录存在
            SESSIONS_FILE.parent.mkdir(parents=True, exist_ok=True)

            with self._lock:
                data = [s.to_dict() for s in self._sessions.values()]

            with open(SESSIONS_FILE, "w", encoding="utf-8") as f:
                json.dump(data, f, ensure_ascii=False, indent=2)
        except Exception as e:
            print(f"保存 sessions 失败: {e}")

    def _get_service(self):
        """延迟加载 AnalysisService"""
        if self._analysis_service is None:
            from .analysis_service import AnalysisService
            self._analysis_service = AnalysisService()
        return self._analysis_service

    def _get_next_agent(self, current_agent: str) -> str:
        """
        获取下一个要运行的 Agent

        Args:
            current_agent: 当前完成的 Agent 名称

        Returns:
            下一个 Agent 名称，如果没有则返回空字符串
        """
        try:
            idx = AGENT_ORDER.index(current_agent)
            if idx + 1 < len(AGENT_ORDER):
                return AGENT_ORDER[idx + 1]
        except ValueError:
            pass
        return ""

    def create_session(
        self,
        ticker: str,
        stock_name: str,
        market: str,
        start_date: str,
        end_date: str,
        user_id: str = "",
    ) -> Session:
        """
        创建新 Session

        Args:
            ticker: 股票代码
            stock_name: 股票名称
            market: 市场类型
            start_date: 开始日期
            end_date: 结束日期
            user_id: 所属用户

        Returns:
            新创建的 Session
        """
        session_id = str(uuid.uuid4())[:8]  # 短 ID
        session = Session(
            id=session_id,
            ticker=ticker,
            stock_name=stock_name,
            market=market,
            start_date=start_date,
            end_date=end_date,
            user_id=user_id,
        )

        with self._lock:
            self._sessions[session_id] = session

        self._save_sessions()
        return session

    def get_session(self, session_id: str, user_id: str = None, is_admin: bool = False) -> Optional[Session]:
        """
        获取 Session

        Args:
            session_id: Session ID
            user_id: 当前用户 ID（用于权限检查）
            is_admin: 是否是管理员

        Returns:
            Session 或 None（无权限或不存在）
        """
        with self._lock:
            session = self._sessions.get(session_id)
            if session is None:
                return None
            # 管理员可以访问所有 session
            if is_admin:
                return session
            # 普通用户只能访问自己的 session
            if user_id and session.user_id and session.user_id != user_id:
                return None
            return session

    def get_all_sessions(self, user_id: str = None, is_admin: bool = False) -> List[Session]:
        """
        获取所有 Session（按创建时间降序）

        Args:
            user_id: 当前用户 ID（用于过滤）
            is_admin: 是否是管理员（可查看所有）

        Returns:
            Session 列表
        """
        with self._lock:
            if is_admin:
                # 管理员可以看到所有 session
                sessions = list(self._sessions.values())
            elif user_id:
                # 普通用户只能看到自己的 session
                sessions = [s for s in self._sessions.values() if s.user_id == user_id]
            else:
                # 没有 user_id 时返回空（向后兼容：也返回没有 user_id 的旧 session）
                sessions = [s for s in self._sessions.values() if not s.user_id]
        return sorted(sessions, key=lambda s: s.created_at, reverse=True)

    def get_running_count(self, user_id: str = None, is_admin: bool = False) -> int:
        """
        获取正在运行的 Session 数量

        Args:
            user_id: 当前用户 ID（用于过滤）
            is_admin: 是否是管理员

        Returns:
            运行中的 session 数量
        """
        with self._lock:
            if is_admin:
                return sum(
                    1 for s in self._sessions.values()
                    if s.status == SessionStatus.RUNNING
                )
            elif user_id:
                return sum(
                    1 for s in self._sessions.values()
                    if s.status == SessionStatus.RUNNING and s.user_id == user_id
                )
            else:
                return sum(
                    1 for s in self._sessions.values()
                    if s.status == SessionStatus.RUNNING and not s.user_id
                )

    def delete_session(self, session_id: str) -> bool:
        """删除 Session"""
        with self._lock:
            if session_id in self._sessions:
                del self._sessions[session_id]
                self._save_sessions()
                return True
            return False

    def start_session(
        self,
        session_id: str,
        selected_analysts: List[str] = None,
        on_update: Callable[[Session], None] = None,
    ):
        """
        启动 Session 执行

        Args:
            session_id: Session ID
            selected_analysts: 启用的分析师列表
            on_update: 状态更新回调
        """
        session = self.get_session(session_id)
        if not session:
            return

        if on_update:
            self._callbacks[session_id] = on_update

        # 提交到线程池执行
        self._executor.submit(
            self._run_session,
            session_id,
            selected_analysts or ["market", "social", "news", "fundamentals"],
        )

    def _run_session(self, session_id: str, selected_analysts: List[str]):
        """
        执行 Session 分析（在后台线程中运行）

        Args:
            session_id: Session ID
            selected_analysts: 启用的分析师列表
        """
        session = self.get_session(session_id)
        if not session:
            return

        # 更新状态为运行中，设置第一个 Agent
        with self._lock:
            session.status = SessionStatus.RUNNING
            session.current_agent = AGENT_ORDER[0] if AGENT_ORDER else ""

        self._notify_update(session_id)

        try:
            service = self._get_service()

            # 流式分析
            for agent_name, content in service.analyze_stream(
                ticker=session.ticker,
                date=session.end_date,  # 使用结束日期作为分析日期
                market=session.market,
                selected_analysts=selected_analysts,
            ):
                if agent_name == "__FINAL__":
                    # 最终决策
                    with self._lock:
                        session.decision = content
                        session.status = SessionStatus.COMPLETED
                        session.current_agent = ""
                    self._notify_update(session_id)
                    return

                elif agent_name == "__QUOTA_ERROR__":
                    with self._lock:
                        session.status = SessionStatus.ERROR
                        session.error_msg = "API 额度已用完"
                        session.current_agent = ""
                    self._notify_update(session_id)
                    return

                elif agent_name == "__TIMEOUT_ERROR__":
                    with self._lock:
                        session.status = SessionStatus.ERROR
                        session.error_msg = "网络响应超时"
                        session.current_agent = ""
                    self._notify_update(session_id)
                    return

                elif agent_name == "__ERROR__":
                    with self._lock:
                        session.status = SessionStatus.ERROR
                        session.error_msg = content[:200]  # 限制错误信息长度
                        session.current_agent = ""
                    self._notify_update(session_id)
                    return

                else:
                    # Agent 输出
                    with self._lock:
                        session.progress.append(agent_name)
                        session.reports[agent_name] = content
                        # 设置为下一个要运行的 Agent
                        session.current_agent = self._get_next_agent(agent_name)
                    self._notify_update(session_id)

        except Exception as e:
            with self._lock:
                session.status = SessionStatus.ERROR
                session.error_msg = str(e)[:200]
                session.current_agent = ""
            self._notify_update(session_id)

    def _notify_update(self, session_id: str):
        """通知状态更新并保存"""
        # 保存到文件
        self._save_sessions()

        # 发布到 SSE
        global _sse_publisher
        if _sse_publisher:
            session = self.get_session(session_id)
            if session:
                try:
                    _sse_publisher(session_id, session.to_dict())
                except Exception:
                    pass  # 忽略 SSE 发布错误

        # 执行回调
        callback = self._callbacks.get(session_id)
        if callback:
            session = self.get_session(session_id)
            if session:
                try:
                    callback(session)
                except Exception:
                    pass  # 忽略回调错误

    def retry_session(
        self,
        session_id: str,
        selected_analysts: List[str] = None,
        on_update: Callable[[Session], None] = None,
    ):
        """
        重试失败的 Session

        Args:
            session_id: Session ID
            selected_analysts: 启用的分析师列表
            on_update: 状态更新回调
        """
        session = self.get_session(session_id)
        if not session:
            return

        # 重置 Session 状态
        with self._lock:
            session.status = SessionStatus.PENDING
            session.error_msg = ""
            session.progress = []
            session.reports = {}
            session.decision = ""
            session.current_agent = ""

        self._save_sessions()

        if on_update:
            self._callbacks[session_id] = on_update

        # 提交到线程池执行
        self._executor.submit(
            self._run_session,
            session_id,
            selected_analysts or ["market", "social", "news", "fundamentals"],
        )

    def clear_completed(self):
        """清除已完成的 Session"""
        with self._lock:
            to_delete = [
                sid for sid, s in self._sessions.items()
                if s.status in (SessionStatus.COMPLETED, SessionStatus.ERROR)
            ]
            for sid in to_delete:
                del self._sessions[sid]
                if sid in self._callbacks:
                    del self._callbacks[sid]

        if to_delete:
            self._save_sessions()

    def shutdown(self):
        """关闭管理器"""
        self._executor.shutdown(wait=False)


# 全局单例
_session_manager: Optional[SessionManager] = None


def get_session_manager() -> SessionManager:
    """获取全局 SessionManager 单例"""
    global _session_manager
    if _session_manager is None:
        _session_manager = SessionManager()
    return _session_manager
